contracts/extensions/erc1155/ERC1155TokenExtension.sol

```diff
@@ -0,0 +1,361 @@
+        } else {
+            erc1155.safeTransferFrom(
+                msg.sender,
+                address(this),
+                nftTokenId,
+                amount,
+                "0x0"
+            );
+            _saveNft(nftAddr, nftTokenId, GUILD, amount);
+
+            emit CollectedNFT(nftAddr, nftTokenId, amount);
+        }
+    }
+
+    /**
+     * @notice Transfers the NFT token from the extension address to the new owner.
+     * @notice It also updates the internal state to keep track of the all the NFTs collected by the extension.
+     * @notice The caller must have the ACL Flag: WITHDRAW_NFT
+     * @dev Reverts if the NFT is not in ERC721 standard.
+     * @param newOwner The address of the new owner.
+     * @param nftAddr The NFT address that must be in ERC721 standard.
+     * @param nftTokenId The NFT token id.
+     */
+    function withdrawNFT(
+        address newOwner,
+        address nftAddr,
+        uint256 nftTokenId,
+        uint256 amount
+    ) public hasExtensionAccess(this, AclFlag.WITHDRAW_NFT) {
+        IERC1155 erc1155 = IERC1155(nftAddr);
+        require(
+            erc1155.balanceOf(address(this), nftTokenId) > 0,
+            "no amount to withdraw"
+        );
+
+        // remove / update the tokenID amount from the extension
+        _nftTracker[GUILD][nftAddr][nftTokenId].remove(amount);
+
+        // Remove the NFT from the contract address to the actual owner
+        erc1155.safeTransferFrom(
+            address(this),
+            newOwner,
+            nftTokenId,
+            amount,
+            "0x0"
+        );
+
+        uint256 ownerTokenIdBalance =
+            erc1155.balanceOf(address(this), nftTokenId);
+        if (ownerTokenIdBalance == 0) {
+            delete _ownership[getNFTId(nftAddr, nftTokenId)];
+            delete _nftTracker[GUILD][nftAddr][nftTokenId];
+            _nfts[nftAddr].remove(nftTokenId);
+            if (_nfts[nftAddr].length() == 0) {
+                _nftAddresses.remove(nftAddr);
+            }
+        }
+
+        emit WithdrawnNFT(nftAddr, nftTokenId, amount, newOwner);
+    }
+
+    /**
+     * @notice Updates internally the ownership of the NFT.
+     * @notice The caller must have the ACL Flag: INTERNAL_TRANSFER
+     * @dev Reverts if the NFT is not already internally owned in the extension.
+     * @param nftAddr The NFT address.
+     * @param nftTokenId The NFT token id.
+     * @param amount the number of a partricular NFT token id.
+     * @param newOwner The address of the new owner.
+     */
+    function internalTransfer(
+        address nftAddr,
+        uint256 nftTokenId,
+        uint256 amount,
+        address newOwner
+    ) public hasExtensionAccess(this, AclFlag.INTERNAL_TRANSFER) {
+        require(newOwner != address(0x0), "new owner is 0");
+        address currentOwner = _ownership[getNFTId(nftAddr, nftTokenId)];
+        require(currentOwner != address(0x0), "nft not found");
+
```

Comment:
I believe we would need to check the balance of the token before moving it. If the balance is 0, there is nothing to moving, right?